File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> set;

        int l = 0;
        int r = 0;
        int maxLength = 0;

        while (r < s.size()) {
            if(set.contains(s[r])) {
                set.erase(s[l]);
                l++;
            } else {
                set.insert(s[r]);
                maxLength = max(maxLength, r - l + 1);
                r++;
            }
        }

        return maxLength;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> set = new HashSet<>();

        int res = 0;
        int l = 0;
        for (int r = 0; r < s.length(); r++) {
            if (set.contains(s.charAt(r))) {
                res = Math.max(res, r - l);
                while (s.charAt(l) != s.charAt(r)) set.remove(s.charAt(l++));
                l++;
            } else {
                set.add(s.charAt(r));
            }
        }

        return Math.max(res, s.length() - l);
    }
}